Dungeoneer/src/com/interrupt/dungeoneer/generator/SectionDefinition.java

```diff
@@ -19,36 +19,114 @@ public SectionDefinition() { }
     /** Number of levels in this section. */
     public int floors = 2;
 
-    /** Transition level to use after all floors in this section. If numFloors is 0, just use this level. */
+    /** Transition level to use after all floors in this section. If `floors` is 0, just use this level. */
     public Level transitionLevel = null;
 
     /** Array of possible level templates to use when making floors. */
     public Array<Level> levelTemplates = null;
 
+    /** Array of level template tiers spawn probabilities. Use indices of the `levelTemplates` array. */
+    public Array<Integer> levelTemplateTiers = null;
+
+    private Random randomGenerator = new Random();
+
+    /** Returns an array of level templates, ordered by floors, transition level last. */
     public Array<Level> buildLevels() {
-        Array<Level> l = new Array<Level>();
+        Array<Level> levels = new Array<>();
+        Level levelTemplate;
+
+        if(hasFloors() && hasLevelTemplates()) {
+            for(int floor = 0; floor < floors; floor++) {
+                levelTemplate = pickLevelTemplate(floor);
+                if (levelTemplate != null) {
+                    levels.add(levelTemplate);
+                }
+            }
+        }
+
+        if (hasTransitionLevel()) {
+            levels.add(pickTransitionLevel());
+        }
 
-        // make the levels for this dungeon section
-        for(int i = 0; i < floors; i++) {
-            if(levelTemplates != null && levelTemplates.size > 0) {
-                Level picked = levelTemplates.first();
-                Level copy = (Level)KryoSerializer.copyObject(picked);
+        return levels;
+    }
 
-                if(copy.levelName == null || copy.levelName.equals(""))
-                    copy.levelName = name + " " + (i + 1);
+    /** Picks a level template tier. */
+    private int pickLevelTemplateTier() {
+        int tier = -1;
 
-                copy.dungeonLevel = difficultyLevel + i;
-                l.add(copy);
+        if (levelTemplateTiers != null && levelTemplateTiers.size > 0) {
+            try {
+                tier = levelTemplateTiers.get(randomGenerator.nextInt(levelTemplateTiers.size));
```

Comment:
The libGDX `Array` object provides a `random()` method which can be used here.

Reply:
That would also require a change of `tier = randomGenerator.nextInt(levelTemplates.size);`, as it uses the same `randomGenerator` variable. In fact the logic would need to be altered substantially.

But I see the advantage of not having a dependency on `Random`.

Reply:
Adjusted. Closing.